tests/configurable_cluster.py

```diff
@@ -0,0 +1,201 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Cluster object used to control a cluster that can be configured with
+a yaml file.
+
+Test writers should use this module for all of their cluster related needs.
+"""
+
+import fnmatch
+import os
+import paramiko
+from subprocess import check_call
+import tempfile
+import yaml
+
+from prestoadmin import main_dir
+from tests.product import user_password
+
+CONFIG_FILE_GLOB = r'*.yaml'
+DEFAULT_MOUNT_POINT = '/mnt/presto-admin'
+DIST_DIR = os.path.join(main_dir, 'tmp/installer')
+
+
+class ConfigurableCluster(object):
+    """Start/stop/control/query a cluster defined by a configuration file.
+
+    This class allows you to run the presto-admin product tests on a real
+    cluster.
+
+    """
+    def __init__(self, config_filename):
+        # see PyDoc for all_internal_hosts() for an explanation on the
+        # difference between an internal and regular host
+        with open(os.path.join(main_dir, config_filename)) as config_file:
+            config = yaml.load(config_file)
+
+        self.master = config['master']
+        self.slaves = config['slaves']
+        self.internal_master = config['internal_master']
+        self.internal_slaves = config['internal_slaves']
+        if self.internal_master != 'master' or \
+                self.internal_slaves != ['slave1', 'slave2', 'slave3']:
+            raise Exception('Must configure cluster so that the internal '
+                            'master and slaves are master and [slave1, slave2,'
+                            ' slave3] respectively. This may require adding'
+                            ' aliases in /etc/hosts.')
+        self.user = config['user']
+        self.password = user_password
+        self.config = config
+        self.mount_dir = DEFAULT_MOUNT_POINT
+
+    @staticmethod
+    def check_for_cluster_config():
+        config_name = fnmatch.filter(os.listdir(main_dir), CONFIG_FILE_GLOB)
+        if config_name:
+            return config_name[0]
+        else:
+            return None
+
+    def all_hosts(self):
+        return self.slaves + [self.master]
+
+    def all_internal_hosts(self):
+        return self.internal_slaves + [self.internal_master]
+
+    def get_dist_dir(self, unique):
+        if unique:
+            return os.path.join(DIST_DIR, self.master)
+        else:
+            return DIST_DIR
+
+    def tear_down(self):
+        for host in self.all_hosts():
+            script = """
+            if [ -e /etc/hosts_bak ];
+            then mv /etc/hosts_bak /etc/hosts;
+            fi
+            service presto stop
+            rpm -e presto
+            rm -rf /opt/prestoadmin
+            rm -rf /etc/opt/prestoadmin
+            rm -rf /var/log/prestoadmin
+            rm -rf /tmp/presto-debug
+            rm -rf %s
+            """ % self.mount_dir
+            self.run_script_on_host(script, host)
+
+    def stop_host(self, host_name):
```

Comment:
This made me realize that we actually basically have 2 methods, stop_host (for tear down) and disconnect_host (for down node tests) that for docker happens to be done in the same way.  Do you think it's worth separating that out? I don't care particularly.